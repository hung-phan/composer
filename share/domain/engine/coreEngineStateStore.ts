import produce, {
  Patch,
  applyPatches,
  enableMapSet,
  enablePatches,
  setAutoFreeze,
} from "immer";
import * as _ from "lodash";
import { DefaultRootState } from "react-redux";
import actionCreatorFactory from "typescript-fsa";
import { reducerWithInitialState } from "typescript-fsa-reducers";
import { asyncFactory } from "typescript-fsa-redux-thunk";

import {
  Element,
  ElementState,
  HttpMethod,
  Id,
  StateHolderElement,
} from "../interfaces";
import { engineDispatch } from "./coreEngine";
import { ROOT_ID } from "./index";

enableMapSet();
enablePatches();
setAutoFreeze(false);

export interface Container {
  parent?: Id;
  element?: Element;
  childs?: Id[];
}

export type State = {
  [key: Id]: Container;
};

export const mountPoint = "templateEngine";

export const selectors = {
  getState: (state: DefaultRootState): State => state[mountPoint],
  getElement: (state: DefaultRootState, id: Id): Element | undefined => {
    const node = state[mountPoint][id];

    if (!node) {
      throw new Error(`Cannot find state for ${id}`);
    }

    return node.element;
  },
  getElementState: <T extends ElementState<any>>(
    state: DefaultRootState,
    id: Id
  ): T | undefined => {
    const element = state[mountPoint][id].element;

    if (element === undefined) {
      return undefined;
    }

    if (!(element instanceof StateHolderElement)) {
      throw new Error("Current element is not a StateHolderElement");
    }

    return element.elementState as T;
  },
};

const actionCreator = actionCreatorFactory(mountPoint);
const asyncActionCreator = asyncFactory<DefaultRootState>(actionCreator);

export enum ActionType {
  CALL_ENDPOINT = "CALL_ENDPOINT",
  SET_ELEMENT = "SET_ELEMENT",
  DEL_ELEMENT = "DEL_ELEMENT",
  REGISTER_PARENT = "REGISTER_PARENT",
  REPLACE_ELEMENT = "REPLACE_ELEMENT",
  REPLACE_ELEMENT_IN_LIST = "REPLACE_ELEMENT_IN_LIST",
  UPDATE_STATE_ELEMENT = "UPDATE_STATE_ELEMENT",
}

export const actions = {
  callEndpoint: asyncActionCreator<HttpMethod<any>, State>(
    ActionType.CALL_ENDPOINT,
    async (method, dispatch, getRootState) => {
      await engineDispatch(dispatch, [method]);

      return selectors.getState(getRootState());
    }
  ),
  setElement: actionCreator<{ element: Element }>(ActionType.SET_ELEMENT),
  delElement: actionCreator<{ id: Id; interfaceName: string }>(
    ActionType.DEL_ELEMENT
  ),
  registerParent: actionCreator<{ ids: Id[]; parentId: Id }>(
    ActionType.REGISTER_PARENT
  ),
  replaceElement: actionCreator<{
    oldId: Id;
    id: Id;
  }>(ActionType.REPLACE_ELEMENT),
  replaceElementInList: actionCreator<{
    oldId: Id;
    ids: Id[];
  }>(ActionType.REPLACE_ELEMENT_IN_LIST),
  updateStateElement: actionCreator<{
    stateElementId: string;
    patches: Patch[];
  }>(ActionType.UPDATE_STATE_ELEMENT),
};

function replaceElement(
  oldId: string,
  element: Element,
  parentElement: Element
): void {
  for (const [key, value] of Object.entries(parentElement)) {
    if (value instanceof Element && value.id === oldId) {
      parentElement[key] = element;
      return;
    } else if (_.isArray(value)) {
      for (let index = 0, len = value.length; index < len; index++) {
        if (value[index] instanceof Element && value[index].id === oldId) {
          value[index] = element;
          return;
        }
      }
    }
  }
}

function replaceElementInList(
  oldId: string,
  elements: Element[],
  parentElement: Element
): void {
  for (const value of Object.values(parentElement)) {
    if (_.isArray(value)) {
      for (let index = 0, len = value.length; index < len; index++) {
        if (value[index] instanceof Element && value[index].id === oldId) {
          value.splice(index, 1, ...elements);

          return;
        }
      }
    }
  }
}

const EMPTY_ARRAY = [];

function getChilds(state: State, id: Id): Id[] {
  return state[id].childs !== undefined ? state[id].childs : EMPTY_ARRAY;
}

function deleteAllChildElements(state: State, id: Id): void {
  if (!(id in state)) {
    return;
  }

  for (const child of getChilds(state, id)) {
    deleteAllChildElements(state, child);
  }

  delete state[id];
}

export const reducer = reducerWithInitialState<State>({})
  .case(actions.setElement, (state, action) =>
    produce(state, (draft) => {
      const element = action.element;

      if (!(element.id in draft)) {
        draft[element.id] = {};
      }

      draft[element.id].element = element;

      const parentId = draft[element.id].parent;

      if (parentId === undefined) {
        return;
      }

      const parentElement = draft[parentId].element;

      replaceElement(
        element.id,
        Element.builder()
          .id(element.id)
          .interfaceName(element.interfaceName)
          .build(),
        parentElement
      );
    })
  )
  .case(actions.delElement, (state, action) =>
    produce(state, (draft) => {
      if (
        action.id in draft &&
        action.interfaceName === draft[action.id].element.interfaceName
      ) {
        let id = action.id;
        let pointer = draft[id];

        while (pointer.parent !== undefined) {
          const currentId = id;

          id = pointer.parent;
          pointer = draft[id];

          if (!_.find(getChilds(draft, id), (child) => child === currentId)) {
            deleteAllChildElements(draft, currentId);
            return;
          }
        }

        if (id !== ROOT_ID) {
          delete draft[action.id];
        }
      }
    })
  )
  .case(actions.registerParent, (state, action) =>
    produce(state, (draft) => {
      if (!(action.parentId in draft)) {
        draft[action.parentId] = {};
      }

      draft[action.parentId].childs = [];

      for (const id of action.ids) {
        if (
          !_.find(getChilds(draft, action.parentId), (child) => child === id)
        ) {
          if (draft[action.parentId].childs === undefined) {
            draft[action.parentId].childs = [];
          }

          draft[action.parentId].childs.push(id);

          if (!(id in draft)) {
            draft[id] = {};
          }

          draft[id].parent = action.parentId;
        }
      }
    })
  )
  .case(actions.replaceElement, (state, action) =>
    produce(state, (draft) => {
      const parentId = draft[action.oldId].parent;

      if (parentId === undefined) {
        return;
      }

      const parentElement = draft[parentId].element;
      const newElement = draft[action.id].element;

      replaceElement(
        action.oldId,
        Element.builder()
          .id(newElement.id)
          .interfaceName(newElement.interfaceName)
          .build(),
        parentElement
      );

      draft[action.oldId].parent = undefined;

      const childs = getChilds(state, parentId);
      if (childs.length > 0) {
        draft[parentId].childs = childs.filter(
          (child) => child !== action.oldId
        );
      }

      draft[action.id].parent = parentId;
      draft[parentId].childs.push(action.id);
    })
  )
  .case(actions.replaceElementInList, (state, action) =>
    produce(state, (draft) => {
      const parentId = draft[action.oldId].parent;

      if (parentId === undefined) {
        return;
      }

      const parentElement = draft[parentId].element;
      const newElements = action.ids.map((id) => draft[id].element);

      replaceElementInList(
        action.oldId,
        newElements.map((newElement) =>
          Element.builder()
            .id(newElement.id)
            .interfaceName(newElement.interfaceName)
            .build()
        ),
        parentElement
      );

      draft[action.oldId].parent = undefined;

      const childs = getChilds(state, parentId);
      if (childs.length > 0) {
        draft[parentId].childs = childs.filter(
          (child) => child !== action.oldId
        );
      }

      for (const id of action.ids) {
        draft[id].parent = parentId;
        draft[parentId].childs.push(id);
      }
    })
  )
  .case(actions.updateStateElement, (state, action) =>
    produce(state, (draft) => {
      const element = draft[action.stateElementId].element;

      if (!(element instanceof StateHolderElement)) {
        throw new Error("Can only update state of StateHolderElement");
      }

      applyPatches(element.elementState, action.patches);
    })
  );
