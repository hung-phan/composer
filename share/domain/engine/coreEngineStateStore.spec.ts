import {
  Element,
  LayoutElement,
  PlaceholderElement,
} from "../interfaces";
import { actions, reducer } from "./coreEngineStateStore";
import { ROOT_ID } from "./index";

test("setElement", () => {
  let state = {};

  state = reducer(
    state,
    actions.setElement({
      element: Element.builder().id(ROOT_ID).build(),
    })
  );
  state = reducer(
    state,
    actions.setElement({
      element: Element.builder().id("element_id_1").build(),
    })
  );
  state = reducer(
    state,
    actions.setElement({
      element: Element.builder()
        .id("element_id_2")
        .interfaceName("PlaceholderElement")
        .build(),
    })
  );

  state = reducer(
    state,
    actions.registerParent({
      ids: ["element_id_1", "element_id_2"],
      parentId: ROOT_ID,
    })
  );

  expect(state).toMatchSnapshot();

  state = reducer(
    state,
    actions.setElement({
      element: Element.builder()
        .id("element_id_2")
        .interfaceName("ButtonElement")
        .build(),
    })
  );

  state = reducer(
    state,
    actions.setElement({
      element: Element.builder().id("element_id_3").build(),
    })
  );

  state = reducer(
    state,
    actions.setElement({
      element: Element.builder().id("element_id_4").build(),
    })
  );

  state = reducer(
    state,
    actions.registerParent({
      ids: ["element_id_3", "element_id_4"],
      parentId: "element_id_2",
    })
  );

  expect(state).toMatchSnapshot();
});

test("delElement", () => {
  expect(
    reducer(
      {
        element_id: {
          element: Element.builder().id("element_id").build(),
        },
      },
      actions.delElement({
        id: "element_id",
        interfaceName: "Element",
      })
    )
  ).toMatchSnapshot();
});

test("delElement that is not connected to parent", () => {
  const state = {
    [ROOT_ID]: {
      element: Element.builder().id("element_id").build(),
      childs: ["child_element_id_1", "child_element_id_2"],
    },

    child_element_id_1: {
      parent: ROOT_ID,
      element: Element.builder().id("child_element_id_1").build(),
    },

    child_element_id_2: {
      parent: ROOT_ID,
      element: Element.builder().id("child_element_id_2").build(),
    },

    child_element_id_3: {
      parent: ROOT_ID,
      element: Element.builder().id("child_element_id_3").build(),
    },
  };

  expect(
    reducer(
      state,
      actions.delElement({
        id: "child_element_id_1",
        interfaceName: "Element",
      })
    )
  ).toMatchSnapshot();

  expect(
    reducer(
      state,
      actions.delElement({
        id: "child_element_id_2",
        interfaceName: "Element",
      })
    )
  ).toMatchSnapshot();

  expect(
    reducer(
      state,
      actions.delElement({
        id: "child_element_id_3",
        interfaceName: "Element",
      })
    )
  ).toMatchSnapshot();
});

describe("template logic", () => {
  let state = {};

  // init state
  test("should update initial state correctly", () => {
    state = reducer(
      state,
      actions.setElement({
        element: Element.builder().id("element_id_1").build(),
      })
    );
    state = reducer(
      state,
      actions.setElement({
        element: PlaceholderElement.builder()
          .id("element_id_2_place_holder")
          .build(),
      })
    );
    state = reducer(
      state,
      actions.setElement({
        element: LayoutElement.builder()
          .id("parent_element_id")
          .elements([
            Element.builder().id("element_id_1").build(),
            Element.builder().id("element_id_2_place_holder").build(),
          ])
          .build(),
      })
    );
    state = reducer(
      state,
      actions.registerParent({
        ids: ["element_id_1", "element_id_2_place_holder"],
        parentId: "parent_element_id",
      })
    );

    expect(state).toMatchSnapshot();
  });

  test("replaceElement should update state correctly", () => {
    state = reducer(
      state,
      actions.setElement({
        element: Element.builder().id("element_id_4").build(),
      })
    );
    state = reducer(
      state,
      actions.setElement({
        element: LayoutElement.builder()
          .id("element_id_3")
          .elements([Element.builder().id("element_id_4").build()])
          .build(),
      })
    );
    state = reducer(
      state,
      actions.registerParent({
        ids: ["element_id_4"],
        parentId: "element_id_3",
      })
    );

    expect(state).toMatchSnapshot();
  });

  test("replaceElement should work correctly", () => {
    state = reducer(
      state,
      actions.replaceElement({
        id: "element_id_3",
        oldId: "element_id_2_place_holder",
      })
    );

    expect(state).toMatchSnapshot();
  });
});
