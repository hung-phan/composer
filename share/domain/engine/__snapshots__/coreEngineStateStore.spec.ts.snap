// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delElement 1`] = `Object {}`;

exports[`delElement that is not connected to parent 1`] = `
Object {
  "@@ENGINE_ROOT_ID": Object {
    "childs": Array [
      "child_element_id_1",
      "child_element_id_2",
    ],
    "element": Element {
      "id": "element_id",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
  },
  "child_element_id_1": Object {
    "element": Element {
      "id": "child_element_id_1",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
  "child_element_id_2": Object {
    "element": Element {
      "id": "child_element_id_2",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
  "child_element_id_3": Object {
    "element": Element {
      "id": "child_element_id_3",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
}
`;

exports[`delElement that is not connected to parent 2`] = `
Object {
  "@@ENGINE_ROOT_ID": Object {
    "childs": Array [
      "child_element_id_1",
      "child_element_id_2",
    ],
    "element": Element {
      "id": "element_id",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
  },
  "child_element_id_1": Object {
    "element": Element {
      "id": "child_element_id_1",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
  "child_element_id_2": Object {
    "element": Element {
      "id": "child_element_id_2",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
  "child_element_id_3": Object {
    "element": Element {
      "id": "child_element_id_3",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
}
`;

exports[`delElement that is not connected to parent 3`] = `
Object {
  "@@ENGINE_ROOT_ID": Object {
    "childs": Array [
      "child_element_id_1",
      "child_element_id_2",
    ],
    "element": Element {
      "id": "element_id",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
  },
  "child_element_id_1": Object {
    "element": Element {
      "id": "child_element_id_1",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
  "child_element_id_2": Object {
    "element": Element {
      "id": "child_element_id_2",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
}
`;

exports[`setElement 1`] = `
Object {
  "@@ENGINE_ROOT_ID": Object {
    "childs": Array [
      "element_id_1",
      "element_id_2",
    ],
    "element": Element {
      "id": "@@ENGINE_ROOT_ID",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
  },
  "element_id_1": Object {
    "element": Element {
      "id": "element_id_1",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
  "element_id_2": Object {
    "element": Element {
      "id": "element_id_2",
      "interfaceName": "PlaceholderElement",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
}
`;

exports[`setElement 2`] = `
Object {
  "@@ENGINE_ROOT_ID": Object {
    "childs": Array [
      "element_id_1",
      "element_id_2",
    ],
    "element": Element {
      "id": "@@ENGINE_ROOT_ID",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
  },
  "element_id_1": Object {
    "element": Element {
      "id": "element_id_1",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
  "element_id_2": Object {
    "childs": Array [
      "element_id_3",
      "element_id_4",
    ],
    "element": Element {
      "id": "element_id_2",
      "interfaceName": "ButtonElement",
      Symbol(immer-draftable): true,
    },
    "parent": "@@ENGINE_ROOT_ID",
  },
  "element_id_3": Object {
    "element": Element {
      "id": "element_id_3",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "element_id_2",
  },
  "element_id_4": Object {
    "element": Element {
      "id": "element_id_4",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "element_id_2",
  },
}
`;

exports[`template logic replaceElement should update state correctly 1`] = `
Object {
  "element_id_1": Object {
    "element": Element {
      "id": "element_id_1",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "parent_element_id",
  },
  "element_id_2_place_holder": Object {
    "element": PlaceholderElement {
      "id": "element_id_2_place_holder",
      "interfaceName": "PlaceholderElement",
      Symbol(immer-draftable): true,
    },
    "parent": "parent_element_id",
  },
  "element_id_3": Object {
    "childs": Array [
      "element_id_4",
    ],
    "element": LayoutElement {
      "elements": Array [
        Element {
          "id": "element_id_4",
          "interfaceName": "Element",
          Symbol(immer-draftable): true,
        },
      ],
      "id": "element_id_3",
      "interfaceName": "LayoutElement",
      Symbol(immer-draftable): true,
    },
  },
  "element_id_4": Object {
    "element": Element {
      "id": "element_id_4",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "element_id_3",
  },
  "parent_element_id": Object {
    "childs": Array [
      "element_id_1",
      "element_id_2_place_holder",
    ],
    "element": LayoutElement {
      "elements": Array [
        Element {
          "id": "element_id_1",
          "interfaceName": "Element",
          Symbol(immer-draftable): true,
        },
        Element {
          "id": "element_id_2_place_holder",
          "interfaceName": "Element",
          Symbol(immer-draftable): true,
        },
      ],
      "id": "parent_element_id",
      "interfaceName": "LayoutElement",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`template logic replaceElement should work correctly 1`] = `
Object {
  "element_id_1": Object {
    "element": Element {
      "id": "element_id_1",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "parent_element_id",
  },
  "element_id_2_place_holder": Object {
    "element": PlaceholderElement {
      "id": "element_id_2_place_holder",
      "interfaceName": "PlaceholderElement",
      Symbol(immer-draftable): true,
    },
    "parent": undefined,
  },
  "element_id_3": Object {
    "childs": Array [
      "element_id_4",
    ],
    "element": LayoutElement {
      "elements": Array [
        Element {
          "id": "element_id_4",
          "interfaceName": "Element",
          Symbol(immer-draftable): true,
        },
      ],
      "id": "element_id_3",
      "interfaceName": "LayoutElement",
      Symbol(immer-draftable): true,
    },
    "parent": "parent_element_id",
  },
  "element_id_4": Object {
    "element": Element {
      "id": "element_id_4",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "element_id_3",
  },
  "parent_element_id": Object {
    "childs": Array [
      "element_id_1",
      "element_id_3",
    ],
    "element": LayoutElement {
      "elements": Array [
        Element {
          "id": "element_id_1",
          "interfaceName": "Element",
          Symbol(immer-draftable): true,
        },
        Element {
          "id": "element_id_3",
          "interfaceName": "LayoutElement",
          Symbol(immer-draftable): true,
        },
      ],
      "id": "parent_element_id",
      "interfaceName": "LayoutElement",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`template logic should update initial state correctly 1`] = `
Object {
  "element_id_1": Object {
    "element": Element {
      "id": "element_id_1",
      "interfaceName": "Element",
      Symbol(immer-draftable): true,
    },
    "parent": "parent_element_id",
  },
  "element_id_2_place_holder": Object {
    "element": PlaceholderElement {
      "id": "element_id_2_place_holder",
      "interfaceName": "PlaceholderElement",
      Symbol(immer-draftable): true,
    },
    "parent": "parent_element_id",
  },
  "parent_element_id": Object {
    "childs": Array [
      "element_id_1",
      "element_id_2_place_holder",
    ],
    "element": LayoutElement {
      "elements": Array [
        Element {
          "id": "element_id_1",
          "interfaceName": "Element",
          Symbol(immer-draftable): true,
        },
        Element {
          "id": "element_id_2_place_holder",
          "interfaceName": "Element",
          Symbol(immer-draftable): true,
        },
      ],
      "id": "parent_element_id",
      "interfaceName": "LayoutElement",
      Symbol(immer-draftable): true,
    },
  },
}
`;
